segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeepertier2/BookkeeperStorage.java

```diff
@@ -0,0 +1,221 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.impl.bookkeepertier2;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.LoggerHelpers;
+import io.pravega.segmentstore.contracts.SegmentProperties;
+import io.pravega.segmentstore.contracts.StreamSegmentNotExistsException;
+import io.pravega.segmentstore.storage.SegmentHandle;
+import io.pravega.segmentstore.storage.Storage;
+import java.io.InputStream;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.curator.framework.CuratorFramework;
+
+/**
+ * Storage adapter for Apache bookkeeper based storage.
+ *
+ * Each segment is represented by a log built on top of bookkeeper ledgers. This implementation follows some of recommendations
+ * here: https://bookkeeper.apache.org/docs/r4.4.0/bookkeeperLedgers2Logs.html
+ * This log is called a StorageLedger. A storage ledger consists of a number of ledgers and metadata for the segment and the ledgers.
+ * The metadata is stored in ZK. It is accessed using the async curator framework.
+ *
+ * Fencing: The recommended implementation of fencing as described in the URL ensures that the latest caller owns the log.
+ * In case of Storage implementation the requirement is different. A Storage with higher epoc is supposed to own the log,
+ * irrespective of the time the fencing happens. Because of this a CAS operation in ZK for a storage ledger decides ownership.
+ *
+ * Here is the algorithm that describes ownership change through the openWrite() call:
+ *
+ * 1. Check the current epoc for the given StorageLedger.
+ * 2. If the current epoc is larger, set it to the new epoc, other wise throw StorageNotPrimary error.
```

Comment:
1. "otherwise"
2. StorageNotPrimaryException (it's not an error)

Reply:
Fixed.